### PR TITLE
Ryan/typescript

### LINKED ISSUE #77: Typescript Support
# **Feature Request**
Basic typings for the ClgImage component and more.

## **Describe the solution you'd like**

Begin moving the project to a Typescript-first experience. This will get harder as the project gets bigger. Happy to help with this transition!

Having prop typings for components is pretty much essential for libraries at this point, as they provide inline documentation.

### PR DESCRIPTION
**Pull request by tanzoniteblack**
# Description

Converts the project into typescript, allowing projects that depend on this to get return types & (possibly) will help prevent contributors like myself from making certain classes of simple errors.

The, dev only, dependencies that are being introduced here are all related to typescript compilation, both for builds & tests. Fortunately, microbundler already has simple support for typescript.

## Issue Ticket Number

Fixes #77 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Fix or improve the documentation (code documentation)
- [ ] This change requires a documentation update

# Checklist

- [x] I have followed the contributing guidelines of this project as mentioned in [CONTRIBUTING.md](/CONTRIBUTING.md)
- [x] I have created an [issue](https://github.com/colbyfayock/next-cloudinary/issues) ticket for this PR
- [x] I have checked to ensure there aren't other open [Pull Requests](https://github.com/colbyfayock/next-cloudinary/pulls) for the same update/change?
- [x] I have performed a self-review of my own code
- [x] I have run tests locally to ensure they all pass
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes needed to the documentation

### CODE REVIEW

**Comment by tanzoniteblack** [comment]
There's a lot more typing that _can_ be done, but most of those will be more for the benefit of this project, not people that are importing this as a dependency. Definitely worth doing, but can be incremental in nature.

File: next-cloudinary/src/lib/cloudinary.ts
Comment (hades200082) [inline next-cloudinary/src/lib/cloudinary.ts:120]: Should `src` be typed as `string` or maybe `URL`?

File: next-cloudinary/src/lib/cloudinary.ts
Comment (hades200082) [inline next-cloudinary/src/lib/cloudinary.ts:35]: Doesn't the Cloudinary SDK have types for `options` and `config` already that we can just plug in here?

File: README.md
Comment (colbyfayock) [inline README.md:181]: can you revert this changE? this will happen automatically via a script after successful merge

File: next-cloudinary/.tsconfig
Comment (colbyfayock) [inline next-cloudinary/.tsconfig:3]: is this specific to the library? should this be Next Cloudinary?

**Comment by colbyfayock** [comment]
thanks @tanzoniteblack and looks like @hades200082 had a few thoughts here (thanks @hades200082) and added some comments myself

File: next-cloudinary/src/lib/cloudinary.ts
Comment (tanzoniteblack) [reply next-cloudinary/src/lib/cloudinary.ts:35]: it 100% obvious to me without a deeper dive if this code was _only_ using the cloudinary config options. The answer is: Probably, but could also wait until this is merged in to be done as an incremental improvement

File: next-cloudinary/.tsconfig
Comment (tanzoniteblack) [reply next-cloudinary/.tsconfig:3]: https://github.com/tsconfig/bases/blob/main/bases/next.json

If json allowed comments, I'd have stuck this in here. This is using the default from next...but the display option isn't actually something that's used by tsc at all, I don't think. So setting it to this was just me specifying where these options are coming from.

File: next-cloudinary/src/lib/cloudinary.ts
Comment (hades200082) [reply next-cloudinary/src/lib/cloudinary.ts:120]: That's a fair comment. Might be worth creating specific issues to tackle specific areas. @colbyfayock what do you think?

File: next-cloudinary/src/lib/cloudinary.ts
Comment (hades200082) [reply next-cloudinary/src/lib/cloudinary.ts:35]: Again, a sensible approach.

The [cloudinary](https://github.com/cloudinary/cloudinary_npm) package has types that might be useful here at some point.

**Review by hades200082 — APPROVED** [review]

**Comment by colbyfayock** [comment]
sounds good i'll merge it in as is and we can open up new issues

@tanzoniteblack @hades200082 would either of you be interested in either opening new issues where you see potential improvements and / or helping out with the adjustments?

i'm still new to the TS world, so while I may eventually get to something like that, not on my near-term radar

otherwise if we have at least issues maybe someone else would be interesting in contributing

**Comment by colbyfayock** [comment]
hey @tanzoniteblack sorry to throw a wrench here, i made a change in the main branch that now imports `package.json` from Next.js for versioning

the issue is that with the changes here on this branch, it's now failing:
```
➜  next-cloudinary git:(ryan/typescript) yarn build
yarn run v1.22.19
warning package.json: No license field
$ microbundle --no-compress --jsx React.createElement --jsxFragment React.Fragment --jsxImportSource react
(rpt2 plugin) Error: /Users/colbyfayock/Code/next-cloudinary/next-cloudinary/src/lib/cloudinary.ts(2,21): semantic error TS2732: Cannot find module 'next/package.json'. Consider using '--resolveJsonModule' to import module with '.json' extension.
```

i tried adding `--resolveJsonModule` to the microbundle statement but that didn't help. any ideas?

i merged in the latest to this branch

**Comment by tanzoniteblack** [comment]
@colbyfayock , where is that `next/package.json` coming from?

```
(base) ryan@ryans-mbp-2:~/Code/OliverSpace/next-cloudinary/next-cloudinary(ryan/typescript⚡)» yarn install
yarn install v1.22.19
warning package.json: No license field
[1/4] 🔍  Resolving packages...
success Already up-to-date.
✨  Done in 0.20s.
(base) ryan@ryans-mbp-2:~/Code/OliverSpace/next-cloudinary/next-cloudinary(ryan/typescript⚡)» find . -name "package.json"
./node_modules/@jest/types/package.json
./node_modules/ts-jest/package.json
./node_modules/typescript/package.json
./node_modules/yargs-parser/package.json
./node_modules/jest-util/package.json
./package.json
(base) ryan@ryans-mbp-2:~/Code/OliverSpace/next-cloudinary/next-cloudinary(ryan/typescript⚡)»
```

**Comment by colbyfayock** [comment]
so im realizing it might be because next.js is only a peer dependency of this project, so i may need to rethink how i'm doing this, I'll revert the other change

**Comment by tanzoniteblack** [comment]
@colbyfayock ,I think I figured it out. Microbundler is looking for `tsconfig.json`, not `.tsconfig`. Give me a few minutes and I think I can have a fix up

**Comment by colbyfayock** [comment]
oof i already reverted it, but if you have a solution would love to hear

**Comment by colbyfayock** [comment]
here was the original PR https://github.com/colbyfayock/next-cloudinary/pull/102

File: next-cloudinary/src/lib/cloudinary.ts
Comment (shairez) [inline next-cloudinary/src/lib/cloudinary.ts:32]: Line 29 - `let cld` should have the type `:Cloudinary`

But when I tried adding that, I got 2 erros about some of the types below (effect not bein a string etc)

So consider checking it out

Comment (colbyfayock) [reply next-cloudinary/src/lib/cloudinary.ts:32]: im wondering if that's because i'm not using those methods as they were designed, i'm passing the transformations directly instead of the imported transformation instances. ill look at this after this PR and see if maybe there's a better method to use

**Comment by colbyfayock** [comment]
@all-contributors please add @tanzoniteblack for code

**Pull request merged**
